openwrt-openwisp-monitoring/files/monitoring.agent

```diff
@@ -78,24 +65,140 @@ collect_data(){
   done
 }
 
-while true
-do
-  data="$(collect_data)"
-  #send data
-  response_code=$($CURL_COMMAND -H "Content-Type: application/json" \
-    -d "$data" \
-    -v "$URL")
-  if [ "$response_code" = "200" ]; then
-    [ "$verbose_mode" = "1" ] && logger -s "Data sent successfully." \
-                                        -p daemon.info
-  else
-    logger -s "Data not sent successfully. Response code is $response_code" \
+set_url_and_curl(){
+  if [ -z "$BASE_URL" ]; then
+    logger -s "missing required --url option" \
            -t openwisp_monitoring \
            -p daemon.err
+    exit 1
+  fi
 
-    [ "$verbose_mode" = "0" ] && logger -s "Run with verbose mode to find more." \
-                                        -t openwisp_monitoring \
-                                        -p daemon.err
+  if [ -z "$UUID" ]; then
+    logger -s "missing required --uuid option" \
+            -t openwisp_monitoring \
+            -p daemon.err
+    exit 1
   fi
-  sleep "$INTERVAL" & wait $!
-done
+
+  if [ -z "$KEY" ]; then
+    logger -s "missing required --key option" \
+           -t openwisp_monitoring \
+           -p daemon.err
+    exit 1
+  fi
+
+  URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
+
+  CURL_COMMAND="curl -s -w "%{http_code}""
+  [ "$VERIFY_SSL" = "0" ] && CURL_COMMAND="$CURL_COMMAND -k"
+  [ "$verbose_mode" = "1" ] && CURL_COMMAND="$CURL_COMMAND -v"
+  echo "0"
+}
+
+save_data() {
+  # Remove double quotes from interfaces
+  MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
+
+  while true
+  do
+    data="$(collect_data)"
```

Comment:
why are you collecting the data before checking for the memory availability?
If the data is large and the memory is already at full capacity, isn't there's a chance the agent my crash here? In that case it would not be as robust as intended wouldn't it?